src/counts.jl

```diff
@@ -450,5 +450,13 @@ Return a dictionary mapping each unique value in `x` to its proportion in `x`.
 If a vector of weights `wv` is provided, the proportion of weights is computed rather
 than the proportion of raw counts.
 """
-proportionmap(x::AbstractArray) = _normalize_countmap(countmap(x), length(x))
+function proportionmap(x)
+    countm = Dict{eltype(x), Int}()
+    n = 0
+    for y in x
+        countm[y] = get(countm, y, 0) + 1
+        n += 1
+    end
```

Comment:
This reinvents `countmap`. Better make `countmap` allow iterators instead, so that both functions benefit.

Reply:
`countmap` already accepts iterators; I did that to keep a count of `n` while iterating.

Reply:
OK. The problem is that `countmap` uses different algorithms under the hood for performance. By using a `Dict` here, you lose the benefit of the fast radix sort and count sort algorithms.

I see two solutions:
- do `n = Base.IteratorSize(x) isa Union{HasLength, HasShape} ? length(x) : sum(values(countm))`
- adjust all `_addcounts!` methods to return the number of elements (this should be cheap so not a big deal if it's not used by `addcounts`)

Reply:
I am looking to help get this across the line. Is this your first proposed solution?

```julia
function proportionmap(x)
    countm = countmap(x)
    n = Base.IteratorSize(x) isa Union{Base.HasLength, Base.HasShape} ? length(x) : sum(values(countm))
    _normalize_countmap(countm, n)
end
```